pkg/scm/git/git.go

```diff
@@ -62,18 +78,178 @@ func stringInSlice(s string, slice []string) bool {
+// since the URL is
+func (h *stiGit) MungeNoProtocolURL(source string, uri *url.URL) error {
+	if uri == nil {
+		return nil
+	}
+
+	justReturn, validURL := schemeNecessitatesReturn(uri)
+	if justReturn {
+		if validURL {
+			return nil
+		} else {
+			errors.New(fmt.Sprintf("this url is not supported for git cloning: %v", uri))
+		}
+	}
+
+	mods, err := parseSpec(source, false)
 	if err != nil {
-		return false
+		return err
+	}
+	// don't modify the protocol options which always have to be explicitly specified (http, https,  git)
+	if uri.Scheme != "ssh" && uri.Scheme != "file" {
+		return nil
 	}
 
-	if stringInSlice(url.Scheme, allowedSchemes) {
-		return true
+	// update the either file or ssh url accordingly
+	if len(mods.user) > 0 {
+		uri.User = url.User(mods.user)
+	}
+	if len(mods.scheme) > 0 {
+		uri.Scheme = mods.scheme
+	}
+	if len(mods.host) > 0 {
+		uri.Host = mods.host
+	}
+	if len(mods.path) > 0 {
+		uri.Path = mods.path
+	}
+	if len(mods.frag) > 0 {
+		uri.Fragment = mods.frag
+	}
+	return nil
+}
+
+// schemeNeccessitatesReturn examines the URL and determines whether further
+// url / git spec string processing is needed or not, and if no more processing is needed,
+// whether the input should be deemed valid or not;
+// the combos of interest are:
+// - if a valid protocol/scheme is specifed, just return (leaving the valid URL alone)
+// - if an invalid protocol/scheme is specified
+// - if no protocol/scheme is specified, don't return and continue trying to construct a valid URL
+func schemeNecessitatesReturn(uri *url.URL) (justReturn, invalidProtocol bool) {
+	if stringInSlice(uri.Scheme, allowedSchemes) {
+		justReturn = true
+		invalidProtocol = false
+		return
+	}
+
+	//at this point, the only valid value for Scheme should be "";  if it is
+	// set to some other protocol besides what is listed in allowedSchemes,
+	// say ftp://, then we return false, as it is a protocol we are not expecting.
+	// note, there also appears to be a bug in url.Parse, where a url like
+	// linux-box:/path/to/code will report the scheme as "linux-box"
+	if uri.Scheme != "" {
+		justReturn = true
+		invalidProtocol = true
+		return
+	}
+
+	justReturn = false
+	invalidProtocol = false
+	return
+}
+
+//func parseSpec(source string, remoteOnly bool) (validURL bool, scheme, user, host, path, ref string) {
+func parseSpec(source string, remoteOnly bool) (*urlMods, error) {
+	uri, err := url.Parse(source)
+	if err != nil {
+		return nil, err
+	}
+
+	if uri.Scheme == "file" && remoteOnly {
+		return nil, errors.New("file protocol is not a remote only spec")
+	}
+
+	justReturn, invalidProtocol := schemeNecessitatesReturn(uri)
+	if invalidProtocol {
+		return nil, errors.New(fmt.Sprintf("unsupported protocol specfied:  %s", uri.Scheme))
+	}
+	if justReturn {
+		return nil, nil
+	}
+
+	lastColonIdx := strings.LastIndex(source, ":")
+	atSignPresent := strings.Index(source, "@") != -1
+	if lastColonIdx != -1 && atSignPresent {
```

Comment:
not to make you feel like i'm going back and forth on you, but this would imply that:
github.com:openshit/origin.git  would not be a valid ssh clone spec, but it appears that it is:
```
$ git clone user@server:project.git

You can also not specify a user, and Git assumes the user you’re currently logged in as.
```
(from https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols)

at the end of the day there may not be any valid way to distinguish between a protocol-less ssh spec, and a protocoless local file spec in which the path includes a colon.

I wonder how git distinguishes, since git does allow local file paths as clone specs too.  Perhaps it just tries the local path first and if it doesn't exist, tries it as an ssh spec.  Presumably we could do the same thing here...  i know you already have some logic that checks if the path provided exists locally, maybe if it does you should just immediately augment the path with file:// (if it doesn't already have that prefix) and then you can avoid a lot of this nastiness.